File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        p1, p2 = 0, len(nums) - 1

        while p1 <= p2:
            pm = (p1 + p2) // 2
            if nums[pm] == target:
                return pm
            elif nums[pm] < target:
                p1 = pm + 1   # target is in the right half
            else:
                p2 = pm - 1   # target is in the left half

        return -1

File: Data Structures & Algorithms/two-integer-sum/submission-1.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        std::unordered_map<int, int> nToIdx;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            nToIdx.insert({target - nums[i], i});
        }

        for (int i = 0; i < n; i++) {
            if (nToIdx.contains(nums[i]) && nToIdx[nums[i]] != i) {
                return {min(i, nToIdx[nums[i]]), max(i, nToIdx[nums[i]])};
            }
        }
        return {};
    }
};
